lib/money/money_parser.rb

```diff
@@ -1,31 +1,41 @@
+# parse a amount from a string
 class MoneyParser
-  # parse a amount from a string
+  MARKS = "\.,·’"
+  EXTRA_MARKS = "\s˙'"
+
   def self.parse(input, currency = nil)
     new.parse(input, currency)
   end
 
   def parse(input, currency = nil)
-    Money.new(extract_money(input.to_s), currency)
+    Money.new(extract_money(input.to_s, currency), currency)
   end
 
   private
-  def extract_money(input)
-    return Money.zero if input.to_s.empty?
 
-    amount = input.scan(/\-?[\d\.\,]+/).first
+  def extract_money(input, currency_iso = nil)
+    return '0' if input.to_s.empty?
+
+    amount = input.scan(/(-?[\d#{MARKS}][\d#{MARKS}#{EXTRA_MARKS}]*)/).first
+    return '0' unless amount
+    amount = amount.first.tr(EXTRA_MARKS, '')
 
-    return Money.zero if amount.nil?
+    *other_marks, last_mark = amount.scan(/[#{MARKS}]/)
+    other_marks.uniq!
 
-    # Convert 0.123 or 0,123 into what will be parsed as a decimal amount 0.12 or 0.13
-    amount.gsub!(/^(-)?(0[,.]\d\d)\d+$/, '\1\2')
+    return amount if last_mark.nil?
 
-    segments = amount.scan(/^(.*?)(?:[\.\,](\d{1,2}))?$/).first
+    *dollars, cents = amount.split(last_mark)
+    dollars = dollars.join.tr(MARKS, '')
 
-    return Money.zero if segments.empty?
+    if cents.size > 2
```

Comment:
Shouldn't we be checking against the currency's number of subunits?

Reply:
This `if` is to check if the cents are actually thousands. No currency represents thousands with less than 3 digits, even currencies with 0 or 3 minor units. In other words it's not possible is the cents has less than 3 digits that it's a thousand amount.